src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -0,0 +1,1194 @@
+                // non-waiters from preempting waiters.
+                //   - When the first waiter begins waiting, it records the current time as a "waiter starvation start time".
+                //     That is a point in time after which no forward progress has occurred for waiters. When a waiter acquires
+                //     the lock, the time is updated to the current time.
+                //   - This method checks whether the starvation duration has crossed a threshold and if so, sets
+                //     ShouldNotPreemptWaitersMask
+                //
+                // When unreasonable starvation is occurring, the lock will be released occasionally and if caused by spinners,
+                // those threads may start to spin again.
+                //   - Before starting to spin this method is called. If ShouldNotPreemptWaitersMask is set, the spinner will
+                //     skip spinning and wait instead. Spinners that are already registered at the time
+                //     ShouldNotPreemptWaitersMask is set will stop spinning as necessary. Eventually, all spinners will drain
+                //     and no new ones will be registered.
+                //   - Upon releasing a lock, if there are no spinners, a waiter will be signaled to wake. On that path,
+                //     TrySetIsWaiterSignaledToWake() is called.
+                //   - Eventually, after spinners have drained, only a waiter will be able to acquire the lock. When a waiter
+                //     acquires the lock, or when the last waiter unregisters itself, ShouldNotPreemptWaitersMask is cleared to
+                //     restore the normal policy.
+
+                Debug.Assert(spinCount >= 0);
+
+                isFirstSpinner = false;
+                var state = new State(lockObj);
+                while (true)
+                {
+                    State newState = state;
+                    TryLockResult result = TryLockResult.Spin;
+                    if (newState.HasAnyWaiters)
+                    {
+                        if (newState.ShouldNotPreemptWaiters)
+                        {
+                            return TryLockResult.Wait;
+                        }
+                        if (lockObj.ShouldStopPreemptingWaiters)
+                        {
+                            newState.SetShouldNotPreemptWaiters();
+                            result = TryLockResult.Wait;
+                        }
+                    }
+                    if (result == TryLockResult.Spin)
+                    {
+                        Debug.Assert(!newState.ShouldNotPreemptWaiters);
+                        if (!newState.IsLocked)
+                        {
+                            newState.SetIsLocked();
+                            result = TryLockResult.Locked;
+                        }
+                        else if ((newState.HasAnySpinners && spinCount == 0) || !newState.TryIncrementSpinnerCount())
```

Comment:
I am not sure tracking the spinner count is useful. Most nonwaiters will acquire the lock via the fast path - without registering as a spinner. We also expect the number of spinners be small. Otherwise we have a very contended lock and blocking is preferrable, since competing spinners are expensive and having a lot of spinners just means that most will not succeed and will have to block anyways.  

Even as a rate-limiting measure for waiter release the number of spinners does not seem very helpful. Since spinning is a way to mask the awake latencies, it is the most effective when spin times are shorter than it takes to wake up a waiter.  Basically - the fact that we wake one waiter at a time will have a lot more rate limiting impact than looking at the presence of spinners.

I think if we do not register/unregister spinners, things will only get simpler.

Reply:
> Otherwise we have a very contended lock and blocking is preferrable, since competing spinners are expensive.

Usually blocking is more expensive. When there's a burst of contention on a lock from a number of threads, typically all of those threads will need to acquire the lock. Blocking eagerly would effectively create an expensive lock convoy. On the other hand, it doesn't seem to help much to have more than 8-16 threads spin-waiting simultaneously even on larger machines at higher contention levels, as many of those spin-waits would fail anyway and we can avoid the extra CPU time.

> Even as a rate-limiting measure for waiter release the number of spinners does not seem very helpful. Since spinning is a way to mask the awake latencies, it is the most effective when spin times are shorter than it takes to wake up a waiter. Basically - the fact that we wake one waiter at a time will have a lot more rate limiting impact than looking at the presence of spinners.

The default max spin count is tuned such that spin times are typically shorter than it takes to block and wake a thread. Waking only one waiter at a time is good, and it's also unnecessary to wake a waiter when there are already spin-waiters that can acquire the lock. Otherwise, there would be a greater chance that a waiter would wake up, fail to acquire the lock, go back to waiting, and repeat. If the woken waiter acquires the lock, there is also a slightly greater chance that it causes a spin-waiter to fail its spin-wait and start waiting, and another waiter would be woken.

Registering spinners seems to work fairly well to enable mild to moderate contention to be resolved with decent latency, while also avoiding some unnecessary waiter wakes and unnecessary spin-waiting under higher contention.